internal/engine/wazevo/backend/isa/amd64/lower_mem.go

```diff
@@ -1,10 +1,116 @@
 package amd64
 
 import (
+	"github.com/tetratelabs/wazero/internal/engine/wazevo/backend"
+	"github.com/tetratelabs/wazero/internal/engine/wazevo/backend/regalloc"
 	"github.com/tetratelabs/wazero/internal/engine/wazevo/ssa"
 )
 
+var addendsMatchOpcodes = [5]ssa.Opcode{ssa.OpcodeUExtend, ssa.OpcodeSExtend, ssa.OpcodeIadd, ssa.OpcodeIconst}
+
 // lowerToAddressMode converts a pointer to an addressMode that can be used as an operand for load/store instructions.
 func (m *machine) lowerToAddressMode(ptr ssa.Value, offsetBase uint32) (am amode) {
-	panic("TODO")
+	offBase := int32(offsetBase)
+	def := m.c.ValueDefinition(ptr)
+	if op := m.c.MatchInstrOneOf(def, addendsMatchOpcodes[:]); op == ssa.OpcodeIadd {
+		x, y := def.Instr.Arg2()
+		xDef, yDef := m.c.ValueDefinition(x), m.c.ValueDefinition(y)
+		rx, offx := m.lowerAddend(xDef)
+		ry, offy := m.lowerAddend(yDef)
+		return m.lowerAddendsToAmode(rx, ry, offx, offy, offBase)
+	} else {
+		// If it is not an Iadd, then we lower the one addend.
+		r, off := m.lowerAddend(def)
+		off += int64(offBase)
+		if r != regalloc.VRegInvalid && lower32willSignExtendTo64(uint64(off)) {
+			return newAmodeImmReg(uint32(off), r)
+		} else {
+			tmpReg := m.c.AllocateVReg(ssa.TypeI64)
+			m.lowerIconst(tmpReg, uint64(off), true)
+			return newAmodeImmReg(0, tmpReg)
+		}
+	}
+}
+
+func (m *machine) lowerAddendsToAmode(rx, ry regalloc.VReg, offx, offy int64, offBase int32) amode {
+	if rx != regalloc.VRegInvalid && offx != 0 || ry != regalloc.VRegInvalid && offy != 0 {
+		panic("invalid input")
+	}
+	u64 := uint64(int64(offBase) + offx + offy)
+	if u64 != 0 && !lower32willSignExtendTo64(u64) {
+		tmpReg := m.c.AllocateVReg(ssa.TypeI64)
+		m.lowerIconst(tmpReg, u64, true)
+		// Blank u64 as it has been already lowered.
+		u64 = 0
+		// We already know that either rx or ry is invalid,
+		// so we overwrite it with the temporary register.
+		if rx == regalloc.VRegInvalid {
+			rx = tmpReg
+		} else {
+			ry = tmpReg
+		}
+	}
+
+	u32 := uint32(u64)
+	switch {
+	// We assume rx, ry are valid iff offx, offy are 0.
+	case rx != regalloc.VRegInvalid && ry != regalloc.VRegInvalid:
+		return newAmodeRegRegShift(u32, rx, ry, 0)
```

Comment:
non zero shift pattern is difficult to implement?

Reply:
doable, it complicates a bit the flow, not sure how worth it is; looking into it.

Reply:
you can write a ssa level pass that counts a patter we are looking for and run it against Zig stdlib on aarch64 maybe :)

Reply:
I added Ishl in b6606bba21abc976ad3981cb1216dbeffa25007f, I can always do the SSA level pass later (that should be fun). We can also revert back this commit if it's too significant.